### PR TITLE
Add order commitment check on verification

### PR DESCRIPTION
**Pull request by fedgiac**
We want to have at most a single valid AMM order on any batch while at the same time providing the flexibility to solvers to create arbitrary orders within the constraints of the AMM function.

This PR enforces this by a commitment mechanism.
If no commitment is set, the verification function assumes that there is an implicit commitment to the order obtained from `getTradeableOrder` and it checks that the input order parameters match.
Any other previously valid order `o` can be settled in a batch as follows.
1. The solver calls `commit(ammAddress, o.hash())` in a pre-interaction.
2. As long as the order passes the AMM checks, signature verification also passes.
3. The solver calls `commit(ammAddress, EMPTY_COMMITMENT)` in a post-interaction to unset the commitment and get back some gas as refund.

It would be really nice to use `TSTORE` to avoid point 3 and to make this process much cheaper but we're a month too early.

Drawbacks of these changes: calling `getTradeableOrder` or committing means an increased gas cost when settling the order.

### To do (in another PR)

Update the docs to explain the role of commitments and their intended usage.

### How to test

Existing and new unit tests. [Here](https://dashboard.tenderly.co/tx/sepolia/0x698a6a232a9cf56ce77909e01630bdb75f128783cdd2166536cc675e10fd957f/gas-usage) is a test order for this code.
You can get a feeling for the gas cost by resimulating without access list. The [simulation](https://dashboard.tenderly.co/fedgiac/one-off-simulations/simulator/8a6adeae-6350-4647-bb69-4f4500f5d72d/gas-usage) highlights that the cost of the extra call to perform the extra checks is about 31k gas.

I also tested that, assuming a commit is set and an order is possible, the watchtower doesn't stop indexing the order. (See internal logs [here](https://production-6de61f.kb.eu-central-1.aws.cloud.es.io/app/r/s/TLDxc).) It does become noisy however.

### CODE REVIEW

**Review by mfw78 — COMMENTED** [review]
One test isn't being executed and a nit on some typos.

File: src/ConstantProduct.sol [hunk 1]
```diff
@@ -259,4 +316,66 @@ contract ConstantProduct is IConditionalOrderGenerator {
...
+    function requireMatchingCommitment(
+        address owner,
+        bytes32 orderHash,
+        bytes calldata staticInput,
+        GPv2Order.Data calldata order
+    ) internal view {
+        bytes32 committedOrderHash = commitment[owner];
+        if (orderHash != committedOrderHash) {
+            if (committedOrderHash != EMPTY_COMMITMENT) {
+                revert IConditionalOrder.OrderNotValid("commitment not matching");
```
[the remark this answers is not in the document]
Comment (fedgiac) [reply src/ConstantProduct.sol:340]: I see. Should I then use dedicated custom errors in the current contract?

File: test/ConstantProduct/DeploymentParamsTest.sol [hunk 2]
```diff
@@ -0,0 +1,13 @@
+// SPDX-License-Identifier: GPL-3.0
+pragma solidity >=0.8.0 <0.9.0;
+
+import {Utils} from "test/libraries/Utils.sol";
+import {ConstantProductTestHarness, ConstantProduct} from "./ConstantProductTestHarness.sol";
+
+abstract contract DeploymentParamsTest is ConstantProductTestHarness {
```
[the remark this answers is not in the document]
Comment (fedgiac) [reply test/ConstantProduct/DeploymentParamsTest.sol:7]: Fixed in 1ea1e651f66181cabda3a1285e1f1b21e634070f.
Im experimenting with making all `ConstantProduct` tests appear under the umbrella of the same test `ConstantProductTest`. Right now this has a few drawbacks, like forgetting to add tests to it, so I'm not sure I'll use this strategy again in the future.

File: src/ConstantProduct.sol (hunk 1 above)
[the remark this answers is not in the document]
Comment (mfw78) [reply src/ConstantProduct.sol:340]: I think in this case it's up to the author. To be strict, I'd choose not to use `OrderNotValid` and use a dedicated custom error, or a simple `require` as you choose :smiley:

[the remark this answers is not in the document]
Comment (fedgiac) [reply src/ConstantProduct.sol]: Updated errors in 2525002f9ab675894b9ad4495f44b710d7e84593.

**Review by mfw78 — APPROVED** [review]

**Pull request merged**